File: lightroom_tagger/core/scoring_service.py
"""Score catalog and Instagram-dump images per perspective into ``image_scores``.

Vision + compression route through :func:`lightroom_tagger.core.vision_op.run_vision_op_persist`
via :func:`lightroom_tagger.core.analyzer.build_score_op_spec`.
"""

from __future__ import annotations

import hashlib
import os
import re
import sqlite3
from collections.abc import Callable
from datetime import datetime, timezone

from lightroom_tagger.core.analyzer import VIDEO_EXTENSIONS, build_score_op_spec
from lightroom_tagger.core.database import (
    get_image,
    get_instagram_dump_media,
    get_perspective_by_slug,
    insert_image_score,
    library_write,
    resolve_filepath,
    supersede_previous_current_scores,
)
from lightroom_tagger.core.prompt_builder import build_scoring_user_prompt
from lightroom_tagger.core.structured_output import StructuredOutputError
from lightroom_tagger.core.vision_cache import get_or_create_cached_image
from lightroom_tagger.core.vision_op import VisionOpOutcome, run_vision_op_persist

LogCallback = Callable[[str, str], None] | None


def compute_prompt_version(perspective_row: dict) -> str:
    """Stable rubric id: ``{slug}:{sha256(prompt_markdown_utf8)[:16]}`` (see D-23)."""
    slug = str(perspective_row["slug"])
    md = str(perspective_row.get("prompt_markdown", "") or "")
    digest = hashlib.sha256(md.encode("utf-8")).hexdigest()[:16]
    return f"{slug}:{digest}"


def perspective_score_already_current(
    conn: sqlite3.Connection,
    image_key: str,
    image_type: str,
    perspective_slug: str,
    prompt_version: str,
) -> bool:
    row = conn.execute(
        """
        SELECT 1 AS ok FROM image_scores
        WHERE image_key = ? AND image_type = ? AND perspective_slug = ?
          AND prompt_version = ? AND is_current = 1
        LIMIT 1
        """,
        (image_key, image_type, perspective_slug, prompt_version),
    ).fetchone()
    return row is not None


def delete_scores_for_version(
    conn: sqlite3.Connection,
    image_key: str,
    image_type: str,
    perspective_slug: str,
    prompt_version: str,
) -> None:
    conn.execute(
        """
        DELETE FROM image_scores
        WHERE image_key = ? AND image_type = ? AND perspective_slug = ?
          AND prompt_version = ?
        """,
        (image_key, image_type, perspective_slug, prompt_version),
    )


def _normalize_perspective_slug(raw_slug: str) -> str:
    return re.sub(r"\s*\(.*\)\s*$", "", raw_slug.strip()).lower().strip()


def _run_score_persist(
    db: sqlite3.Connection,
    *,
    image_key: str,
    image_type: str,
    perspective_slug: str,
    prompt_version: str,
    force: bool,
    image_path: str,
    provider_id: str | None,
    model: str | None,
    log_callback: LogCallback,
    user_prompt: str,
    silent_compression: bool,
) -> VisionOpOutcome:
    spec = build_score_op_spec(
        image_path,
        user_prompt=user_prompt,
        provider_id=provider_id,
        model=model,
        log_callback=log_callback,
        silent_compression=silent_compression,
    )

    reject_reason: dict[str, str] = {}

    def accept_result(parsed_bundle: tuple) -> bool:
        parsed, _repaired = parsed_bundle
        got_slug = _normalize_perspective_slug(parsed.perspective_slug)
        if got_slug != perspective_slug.strip():
            if log_callback:
                log_callback(
                    "warning",
                    f"Slug mismatch: model returned {parsed.perspective_slug!r}, "
                    f"normalized to {got_slug!r}, expected {perspective_slug!r} — skipping",
                )
            reject_reason["msg"] = (
                f"Model returned perspective_slug {parsed.perspective_slug!r}; "
                f"expected {perspective_slug!r}"
            )
            return False
        return True

    def persist(parsed_bundle: tuple, provider: str, model_used: str) -> None:
        parsed, repaired = parsed_bundle
        model_label = f"{provider}:{model_used}"
        scored_at = datetime.now(timezone.utc).replace(microsecond=0).isoformat()
        with library_write(db, log=log_callback):
            if force:
                delete_scores_for_version(
                    db, image_key, image_type, perspective_slug, prompt_version,
                )
            supersede_previous_current_scores(
                db, image_key, image_type, perspective_slug, prompt_version,
            )
            insert_image_score(
                db,
                {
                    "image_key": image_key,
                    "image_type": image_type,
                    "perspective_slug": perspective_slug,
                    "score": parsed.score,
                    "rationale": parsed.rationale,
                    "model_used": model_label,
                    "prompt_version": prompt_version,
                    "scored_at": scored_at,
                    "is_current": 1,
                    "repaired_from_malformed": 1 if repaired else 0,
                    "not_attempted": 1 if parsed.not_attempted else 0,
                },
            )

    outcome = run_vision_op_persist(
        spec,
        accept_result=accept_result,
        persist=persist,
    )
    if outcome.status == "failed" and outcome.reason == "invalid result":
        if "msg" in reject_reason:
            return VisionOpOutcome(status="failed", reason=reject_reason["msg"])
        return VisionOpOutcome(
            status="failed",
            reason="model returned empty or invalid score response",
        )
    return outcome


def score_image_for_perspective(
    db: sqlite3.Connection,
    *,
    image_key: str,
    image_type: str,
    perspective_slug: str,
    force: bool,
    provider_id: str | None,
    model: str | None,
    log_callback: LogCallback = None,
) -> VisionOpOutcome:
    """Score one image for one perspective.

    Returns a :class:`VisionOpOutcome`. ``wrote`` is True when a new score row
    was written.
    """
    if image_type == "catalog":
        image = get_image(db, image_key)
        if not image or not image.get("filepath"):
            return VisionOpOutcome(status="skipped", reason="Catalog image missing or has no filepath")
        filepath = resolve_filepath(str(image["filepath"]))
        if not os.path.exists(filepath):
            return VisionOpOutcome(status="skipped", reason=f"Image file not found: {filepath}")
    elif image_type == "instagram":
        dump = get_instagram_dump_media(db, image_key)
        if not dump or not dump.get("file_path"):
            return VisionOpOutcome(status="skipped", reason="Instagram media missing or has no file_path")
        filepath = str(dump["file_path"])
        if not os.path.exists(filepath):
            return VisionOpOutcome(status="skipped", reason=f"Image file not found: {filepath}")
    else:
        return VisionOpOutcome(status="failed", reason=f"Invalid image_type: {image_type!r}")

    if os.path.splitext(filepath)[1].lower() in VIDEO_EXTENSIONS:
        return VisionOpOutcome(
            status="skipped",
            reason=f"Video file not scorable: {os.path.basename(filepath)}",
        )

    prow = get_perspective_by_slug(db, perspective_slug)
    if not prow:
        return VisionOpOutcome(status="failed", reason=f"Unknown perspective slug: {perspective_slug!r}")
    if int(prow.get("active") or 0) != 1:
        return VisionOpOutcome(status="failed", reason=f"Perspective {perspective_slug!r} is not active")

    pv = compute_prompt_version(prow)

    if not force and perspective_score_already_current(db, image_key, image_type, perspective_slug, pv):
        return VisionOpOutcome(
            status="skipped",
            reason="Score already current for this image, perspective, and prompt version",
        )

    silent_compression = False
    if image_type == "catalog":
        cached = get_or_create_cached_image(db, image_key, filepath)
        if cached and os.path.exists(cached):
            image_for_score = cached
            silent_compression = True
        else:
            image_for_score = filepath
    else:
        image_for_score = filepath

    user_prompt = build_scoring_user_prompt(prow)

    try:
        return _run_score_persist(
            db,
            image_key=image_key,
            image_type=image_type,
            perspective_slug=perspective_slug,
            prompt_version=pv,
            force=force,
            image_path=image_for_score,
            provider_id=provider_id,
            model=model,
            log_callback=log_callback,
            user_prompt=user_prompt,
            silent_compression=silent_compression,
        )
    except StructuredOutputError as exc:
        return VisionOpOutcome(status="failed", reason=str(exc))
    except sqlite3.IntegrityError as exc:
        if "UNIQUE constraint" in str(exc):
            return VisionOpOutcome(status="skipped", reason="Score already written by concurrent worker")
        return VisionOpOutcome(status="failed", reason=str(exc))
    except Exception as exc:
        return VisionOpOutcome(status="failed", reason=str(exc))


__all__ = [
    "compute_prompt_version",
    "delete_scores_for_version",
    "perspective_score_already_current",
    "score_image_for_perspective",
]

File: apps/visualizer/backend-ts/src/vision/scoring-service.ts
/**
 * Score catalog images per perspective into `image_scores`.
 * Port of `core/scoring_service.py`.
 *
 * The describe twin of this module writes one row per image; this one writes one
 * row per image *and rubric version*, which is why so much of it is bookkeeping.
 * A perspective's markdown is editable, and a score means nothing without knowing
 * which wording produced it — so the rubric is hashed into `prompt_version`, and
 * a score is only "already done" when the hash still matches.
 */
import { extname, basename } from 'node:path';
import { createHash } from 'node:crypto';
import { buildScoreOpSpec } from '../analyzer/scoring.js';
import { buildScoringUserPrompt } from '../analyzer/prompt-builder.js';
import type { Db } from '../db/connection.js';
import { getImage } from '../db/library/catalog.js';
import {
  getFrameSubstanceVerdict,
  hasFrameSubstanceOverride,
} from '../db/library/frame-substance.js';
import {
  deleteScoresForVersion,
  getPerspectiveBySlug,
  insertImageScore,
  perspectiveScoreAlreadyCurrent,
  supersedePreviousCurrentScores,
  type PerspectiveRow,
} from '../db/library/scores.js';
import { libraryWrite } from '../db/library/write.js';
import { VIDEO_EXTENSIONS } from '../imaging/raw-decode.js';
import type { ConsecutiveAbortTracker, ErrorPolicy } from '../providers/error-policy.js';
import type { ProviderRegistry } from '../providers/registry.js';
import type { CancelCheck, LogCallback } from '../providers/retry.js';
import { nowIsoUtcSeconds } from '../utils/datetime.js';
import { resolveFilepath } from '../utils/path-resolve.js';
import { StructuredOutputError } from './structured-output.js';
import { resolveVisionImage } from './vision-cache.js';
import { runVisionOpPersist, VisionOpOutcome } from './vision-op.js';

/** Stable rubric id: `{slug}:{sha256(prompt_markdown)[:16]}`. See D-23. */
export function computePromptVersion(row: Pick<PerspectiveRow, 'slug' | 'prompt_markdown'>): string {
  const digest = createHash('sha256')
    .update(row.prompt_markdown ?? '', 'utf8')
    .digest('hex')
    .slice(0, 16);
  return `${String(row.slug)}:${digest}`;
}

/** Strip a trailing parenthetical and case-fold, to compare the model's echoed slug. */
function normalizePerspectiveSlug(rawSlug: string): string {
  return rawSlug
    .trim()
    .replace(/\s*\(.*\)\s*$/, '')
    .toLowerCase()
    .trim();
}

export interface ScoreOptions {
  imageType?: string;
  force?: boolean;
  providerId?: string | null;
  model?: string | null;
  logCallback?: LogCallback;
  registry?: ProviderRegistry | null;
  errorPolicy?: ErrorPolicy | null;
  abortTracker?: ConsecutiveAbortTracker | null;
  cancelCheck?: CancelCheck;
}

/**
 * Score one image for one perspective.
 *
 * `wrote` is true only when a new `image_scores` row was written; every other
 * outcome leaves the table untouched.
 */
export async function scoreImageForPerspective(
  db: Db,
  imageKey: string,
  perspectiveSlug: string,
  opts: ScoreOptions = {},
): Promise<VisionOpOutcome> {
  const imageType = opts.imageType ?? 'catalog';
  const force = Boolean(opts.force);

  if (imageType !== 'catalog') {
    return new VisionOpOutcome('failed', `Invalid image_type: '${imageType}'`);
  }

  const image = getImage(db, imageKey);
  const rawPath = image?.['filepath'];
  if (!image || typeof rawPath !== 'string' || !rawPath) {
    return new VisionOpOutcome('skipped', 'Catalog image missing or has no filepath');
  }
  // Existence is settled later by `resolveVisionImage`, so an unreachable NAS
  // still scores off the local cache.
  const filepath = resolveFilepath(rawPath);

  if (VIDEO_EXTENSIONS.has(extname(filepath).toLowerCase())) {
    return new VisionOpOutcome('skipped', `Video file not scorable: ${basename(filepath)}`);
  }

  // A frame the detector called void has no content to judge. Scoring it anyway
  // would put a number on an empty frame and let it compete in the ranking — an
  // explicit human override is the only thing that reinstates it.
  const substance = getFrameSubstanceVerdict(db, imageKey);
  if (substance?.verdict === 'void' && !hasFrameSubstanceOverride(db, imageKey)) {
    return new VisionOpOutcome('skipped', 'Frame substance verdict: void');
  }

  const perspective = getPerspectiveBySlug(db, perspectiveSlug);
  if (!perspective) {
    return new VisionOpOutcome('failed', `Unknown perspective slug: '${perspectiveSlug}'`);
  }
  if (Number(perspective.active ?? 0) !== 1) {
    return new VisionOpOutcome('failed', `Perspective '${perspectiveSlug}' is not active`);
  }

  const promptVersion = computePromptVersion(perspective);

  if (!force && perspectiveScoreAlreadyCurrent(db, imageKey, imageType, perspectiveSlug, promptVersion)) {
    return new VisionOpOutcome(
      'skipped',
      'Score already current for this image, perspective, and prompt version',
    );
  }

  const { path: imageForScore, silentCompression } = await resolveVisionImage(db, imageKey, filepath);
  if (imageForScore === null) {
    return new VisionOpOutcome('skipped', `Image file not found: ${filepath}`);
  }

  const spec = buildScoreOpSpec(imageForScore, {
    userPrompt: buildScoringUserPrompt(perspective),
    providerId: opts.providerId ?? null,
    model: opts.model ?? null,
    logCallback: opts.logCallback ?? null,
    silentCompression,
    registry: opts.registry ?? null,
    cancelCheck: opts.cancelCheck ?? null,
  });
  spec.errorPolicy = opts.errorPolicy ?? null;
  spec.abortTracker = opts.abortTracker ?? null;

  try {
    const outcome = await runVisionOpPersist(spec, {
      acceptResult: ({ parsed }) => {
        // The score is persisted under the *requested* perspective whatever the
        // model echoes back: a display name that diverges from its slug makes
        // some models paraphrase it, and discarding a valid score over that would
        // cost a provider call to gain nothing. The mismatch is logged instead.
        const got = normalizePerspectiveSlug(parsed.perspective_slug);
        if (got !== perspectiveSlug.trim() && opts.logCallback) {
          opts.logCallback(
            'warning',
            `Slug mismatch: model returned '${parsed.perspective_slug}', normalized to ` +
              `'${got}', expected '${perspectiveSlug}' — persisting under '${perspectiveSlug}'`,
          );
        }
        return true;
      },
      persist: ({ parsed, repaired }, provider, modelUsed) => {
        const scoredAt = nowIsoUtcSeconds();
        libraryWrite(
          db,
          () => {
            if (force) {
              deleteScoresForVersion(db, imageKey, imageType, perspectiveSlug, promptVersion);
            }
            supersedePreviousCurrentScores(db, imageKey, imageType, perspectiveSlug, promptVersion);
            insertImageScore(db, {
              image_key: imageKey,
              image_type: imageType,
              perspective_slug: perspectiveSlug,
              score: parsed.score,
              rationale: parsed.rationale,
              model_used: `${provider}:${modelUsed}`,
              prompt_version: promptVersion,
              scored_at: scoredAt,
              is_current: 1,
              repaired_from_malformed: repaired ? 1 : 0,
              not_attempted: parsed.not_attempted ? 1 : 0,
            });
          },
          { log: opts.logCallback ?? undefined },
        );
      },
    });

    if (outcome.status === 'failed' && outcome.reason === 'invalid result') {
      return new VisionOpOutcome('failed', 'model returned empty or invalid score response');
    }
    return outcome;
  } catch (e) {
    if (e instanceof StructuredOutputError) {
      return new VisionOpOutcome('failed', e.describe());
    }
    const message = e instanceof Error ? e.message : String(e);
    // Two workers can pick the same image and perspective; the unique index is
    // what settles it, and the loser has nothing to report but "already done".
    if (message.includes('UNIQUE constraint')) {
      return new VisionOpOutcome('skipped', 'Score already written by concurrent worker');
    }
    return new VisionOpOutcome('failed', message);
  }
}
